node/core/pvf/src/prepare/worker.rs

```diff
@@ -233,34 +303,79 @@ async fn recv_request(stream: &mut UnixStream) -> io::Result<(Vec<u8>, PathBuf)>
 			"prepare pvf recv_request: non utf-8 artifact path".to_string(),
 		)
 	})?;
-	Ok((code, tmp_file))
+	let preparation_timeout = framed_recv(stream).await?;
+	let preparation_timeout = Duration::decode(&mut &preparation_timeout[..]).map_err(|_| {
+		io::Error::new(
+			io::ErrorKind::Other,
+			"prepare pvf recv_request: failed to decode duration".to_string(),
+		)
+	})?;
+	Ok((code, tmp_file, preparation_timeout))
 }
 
 /// The entrypoint that the spawned prepare worker should start with. The `socket_path` specifies
 /// the path to the socket used to communicate with the host.
 pub fn worker_entrypoint(socket_path: &str) {
 	worker_event_loop("prepare", socket_path, |mut stream| async move {
 		loop {
-			let (code, dest) = recv_request(&mut stream).await?;
+			let (code, dest, preparation_timeout) = recv_request(&mut stream).await?;
 
 			gum::debug!(
 				target: LOG_TARGET,
 				worker_pid = %std::process::id(),
 				"worker: preparing artifact",
 			);
 
+			// Create a lock flag. We set it when either thread finishes.
+			let lock = Arc::new(AtomicBool::new(false));
+
+			let cpu_time_start = ProcessTime::now();
+
+			// Spawn a new thread that runs the CPU time monitor. Continuously wakes up from
+			// sleeping and then either sleeps for the remaining CPU time, or kills the process if
+			// we exceed the CPU timeout.
+			let (stream_2, cpu_time_start_2, preparation_timeout_2, lock_2) =
+				(stream.clone(), cpu_time_start, preparation_timeout, lock.clone());
+			std::thread::Builder::new().name("CPU time monitor".into()).spawn(move || {
+				task::block_on(async {
+					cpu_time_monitor_loop(
+						stream_2,
+						cpu_time_start_2,
+						preparation_timeout_2,
+						lock_2,
+					)
+					.await;
+				})
+			})?;
+
+			// Prepares the artifact in a separate thread.
 			let result = match prepare_artifact(&code) {
 				Err(err) => {
 					// Serialized error will be written into the socket.
 					Err(err)
 				},
 				Ok(compiled_artifact) => {
+					let cpu_time_elapsed = cpu_time_start.elapsed();
+
+					let result =
+						lock.compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst);
+					if result.is_err() {
+						unreachable!(
+							"Hit the timed-out case first; \
+								  Set the flag to true; \
+								  Called process::exit(); \
+								  process::exit: 'This function will never return and \
+								  will immediately terminate the current process.'; \
+								  This is unreachable; qed"
+						);
```

Comment:
This held true when you had a mutex, now it's not. Suppose a sleeping thread wakes up and detects a timeout, it stores `true` in the atomic flag and OS scheduler puts it to a short sleep giving a window for execution of this part of code.

Previously you were holding a locked mutex here, now you don't and there's a panic (although harmless).
Instead, you should join here on a thread handle, giving it time to send error through the socket, then return.

Reply:
You're right, I missed that. I tried to rework this part, let me know how it looks.